google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriterTest.java

```diff
@@ -0,0 +1,601 @@
+    }
+    writer.close();
+  }
+
+  @Test
+  public void testSingleAppendComplexJson() throws Exception {
+    ComplexRoot expectedProto =
+        ComplexRoot.newBuilder()
+            .setTestInt(1)
+            .addTestString("a")
+            .addTestString("b")
+            .addTestString("c")
+            .setTestBytes(ByteString.copyFrom("hello".getBytes()))
+            .setTestBool(true)
+            .addTestDouble(1.1)
+            .addTestDouble(2.2)
+            .addTestDouble(3.3)
+            .addTestDouble(4.4)
+            .setTestDate(1)
+            .setComplexLvl1(
+                com.google.cloud.bigquery.storage.test.JsonTest.ComplexLvl1.newBuilder()
+                    .setTestInt(2)
+                    .setComplexLvl2(
+                        com.google.cloud.bigquery.storage.test.JsonTest.ComplexLvl2.newBuilder()
+                            .setTestInt(3)
+                            .build())
+                    .build())
+            .setComplexLvl2(
+                com.google.cloud.bigquery.storage.test.JsonTest.ComplexLvl2.newBuilder()
+                    .setTestInt(3)
+                    .build())
+            .build();
+    JSONObject complex_lvl2 = new JSONObject();
+    complex_lvl2.put("test_int", 3);
+
+    JSONObject complex_lvl1 = new JSONObject();
+    complex_lvl1.put("test_int", 2);
+    complex_lvl1.put("complex_lvl2", complex_lvl2);
+
+    JSONObject json = new JSONObject();
+    json.put("test_int", 1);
+    json.put("test_string", new JSONArray(new String[] {"a", "b", "c"}));
+    json.put("test_bytes", "hello");
+    json.put("test_bool", true);
+    json.put("test_DOUBLe", new JSONArray(new Double[] {1.1, 2.2, 3.3, 4.4}));
+    json.put("test_date", 1);
+    json.put("complex_lvl1", complex_lvl1);
+    json.put("complex_lvl2", complex_lvl2);
+    JSONArray jsonArr = new JSONArray();
+    jsonArr.put(json);
+
+    JsonStreamWriter writer =
+        getTestJsonStreamWriterBuilder(TEST_STREAM, COMPLEX_TABLE_SCHEMA).build();
+    testBigQueryWrite.addResponse(Storage.AppendRowsResponse.newBuilder().setOffset(0).build());
+    ApiFuture<AppendRowsResponse> appendFuture =
+        writer.append(jsonArr, -1, /* allowUnknownFields */ false);
+
+    assertEquals(0L, appendFuture.get().getOffset());
+    assertEquals(
+        1,
+        testBigQueryWrite
+            .getAppendRequests()
+            .get(0)
+            .getProtoRows()
+            .getRows()
+            .getSerializedRowsCount());
+    assertEquals(
+        testBigQueryWrite.getAppendRequests().get(0).getProtoRows().getRows().getSerializedRows(0),
+        expectedProto.toByteString());
+    writer.close();
+  }
+
+  private final OnSchemaUpdateRunnable ON_SCHEMA_UPDATE_RUNNABLE =
+      new OnSchemaUpdateRunnable() {
+        public void run() {
+          this.getStreamWriter().setUpdatedSchema(this.getUpdatedSchema());
+          try {
+            this.getJsonStreamWriter().setDescriptor(this.getUpdatedSchema());
```

Comment:
Move this definition in test to the JsonWriter. Make sure the connection will always be consistent with the json schema.

Reply:
Got it.